import grpc
import pytest
import requests
import time
import protos.message_pb2 as message_pb2
import protos.message_pb2_grpc as message_pb2_grpc

# Endpoints e portas do servi√ßo
GRPC_SERVER_ADDR = "localhost:50054"
METRICS_URL = "http://localhost:50053"

@pytest.fixture(scope="session")
def grpc_stub():
    """Cria o canal e stub para comunica√ß√£o gRPC com o MessageService."""
    channel = grpc.insecure_channel(GRPC_SERVER_ADDR)
    stub = message_pb2_grpc.MessageServiceStub(channel)
    # Aguarda o servidor iniciar
    for _ in range(10):
        try:
            grpc.channel_ready_future(channel).result(timeout=2)
            print("‚úÖ gRPC MessageService conectado com sucesso!")
            return stub
        except grpc.FutureTimeoutError:
            print("‚è≥ Aguardando servidor gRPC iniciar...")
            time.sleep(2)
    pytest.fail("‚ùå Servidor gRPC (MessageService) n√£o dispon√≠vel ap√≥s v√°rias tentativas.")


def test_send_message_response(grpc_stub):
    """Testa o envio de uma mensagem via gRPC e valida o retorno."""
    request = message_pb2.SendMessageRequest(
        id=1,
        sender="Vitor",
        content="Mensagem teste via gRPC"
    )
    response = grpc_stub.SendMessage(request)
    assert response.id == 1
    assert response.sender == "Vitor"
    assert "Mensagem teste" in response.content
    assert "T" in response.timestamp  # formato ISO8601
    print(f"üì° Resposta gRPC recebida: {response}")


def test_multiple_requests_latency(grpc_stub):
    """Executa m√∫ltiplas chamadas gRPC para medir lat√™ncia e estabilidade."""
    start = time.time()
    for i in range(5):
        req = message_pb2.SendMessageRequest(id=i, sender="User", content=f"Msg {i}")
        resp = grpc_stub.SendMessage(req)
        assert resp.sender == "User"
    elapsed = time.time() - start
    print(f"‚è±Ô∏è 5 requisi√ß√µes gRPC executadas em {elapsed:.3f}s")
    assert elapsed < 5.0  # tempo aceit√°vel total


def test_metrics_exposed():
    """Verifica se o endpoint de m√©tricas Prometheus est√° respondendo corretamente."""
    for _ in range(10):
        try:
            r = requests.get(METRICS_URL)
            if r.status_code == 200:
                content = r.text
                assert "grpc_requests_total" in content
                assert "grpc_request_latency_seconds" in content
                print("üìä M√©tricas gRPC expostas com sucesso!")
                return
        except requests.RequestException:
            pass
        time.sleep(1)
    pytest.fail("‚ùå M√©tricas Prometheus n√£o dispon√≠veis em tempo h√°bil.")
